core/interfaces/i_procedure_model.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * The interface for the data model of a procedure.
+ *
+ * @namespace Blockly.IProcedureModel
+ */
+import * as goog from '../../closure/goog/goog.js';
+goog.declareModuleId('Blockly.IProcedureModel');
+
+import {IParameterModel} from './i_parameter_model.js';
+
+/**
+ * A data model for a procedure.
+ */
+export interface IProcedureModel {
+  /** Sets the human-readable name of the procedure. */
+  setName(name: string): IProcedureModel;
+
+  /**
+   * Inserts a parameter into the list of parameters.
+   *
+   * To move a parameter, first delete it, and then re-insert.
+   */
+  insertParameter(parameterModel: IParameterModel, index: number):
+      IParameterModel;
+
+  /** Removes the parameter at the given index from the parameter list. */
+  deleteParameter(index: number): IProcedureModel;
```

Comment:
Might it be safer for this to take the actual parameter object to be deleted?  Admittedly the implementation might then need to do an array scan internally, but it could avoid issues with unexpectedly interleaved updates to that array.

Reply:
I would rather force people to do an array scan externally if they feel the need, rather than performing one internally every time we do an operation.

Reply:
I guess that's reasonable enough.  And the ordinal position of a parameter is (in most languages, at least) one of its most fundamental properties.